include/sycl/algorithm/sort.hpp

```diff
@@ -173,6 +173,787 @@ void sequential_sort(cl::sycl::queue q, cl::sycl::buffer<T, 1, Alloc> buf,
+    VECTOR_SORT(input2, mDir);
+
+    /* Store the result to global memory */
+    m_globalBuf[global_start + item.get_local_id(0)] = input1;
+    m_globalBuf[global_start + item.get_local_id(0) + 1] = input2;
+  }
+
+private:
+  typename bitonic_sort_base<T, U>::global_buffer_accessor_t m_globalBuf;
+  typename bitonic_sort_base<T, U>::local_buffer_accessor_t m_localBuf;
+  const int mDir;
+};
+
+template <typename T, typename Alloc> class kernel_bitonic_sort_init;
+template <typename T, typename Alloc> class kernel_bitonic_sort_phase_stage_n;
+template <typename T, typename Alloc> class kernel_bitonic_sort_stage_0;
+template <typename T, typename Alloc> class kernel_bitonic_sort_merge;
+template <typename T, typename Alloc> class kernel_bitonic_sort_merge_last;
+
+/* bitonic_sort.
+ * Performs a bitonic sort on the given buffer
+ */
+template <typename T, typename Alloc>
+void bitonic_sort(cl::sycl::queue q, cl::sycl::buffer<T, 1, Alloc> buf,
+                  size_t vectorSize) {
+  using namespace cl::sycl;
+  int direction = 0 /*0 = ascending, -1 = descending*/;
+
+  if (impl::isPowerOfTwo(buf.get_count()) == false) {
+    throw std::runtime_error("Buffer size must be a power-of-two");
+  }
+
+  if (buf.get_count() < 8) {
+    // .. because the bitonic sort allotes 8 elements per work item.
+    throw std::runtime_error("Buffer size must be at least 8");
+  }
+
+  const std::size_t maxWorkGroupSize =
+      q.get_device().get_info<info::device::max_work_group_size>();
+  const std::size_t maxLocalMem =
+      q.get_device().get_info<info::device::local_mem_size>();
+  std::size_t localWorkSize = 2 << static_cast<int>(cl::sycl::log2(
+                                  static_cast<float>(maxWorkGroupSize)));
+
+  const std::size_t localMemPerWorkitem =
+      sizeof(T) * bitonic_sort_base<T, 4>::data_elems_per_thread_::value;
+  std::size_t consumedLocalMem = localWorkSize * localMemPerWorkitem;
+
+  while (consumedLocalMem >= maxLocalMem) {
+    localWorkSize /= 2;
+    consumedLocalMem = localWorkSize * localMemPerWorkitem;
+  }
+
+  const std::size_t globalWorkSize =
+      buf.get_count() / bitonic_sort_base<T, 4>::data_elems_per_thread_::value;
+  if (globalWorkSize < localWorkSize) {
+    localWorkSize = globalWorkSize;
+  }
+
+  range<1> reinterpretedRange(buf.get_count() / vec<T, 4>().get_count());
+  auto reinterpBuf = buf.template reinterpret<vec<T, 4>, 1>(reinterpretedRange);
+
+  auto ndrange = nd_range<1>(range<1>(globalWorkSize), range<1>(localWorkSize));
+
+  q.submit([&](handler &cgh) {
+     auto g = reinterpBuf.template get_access<access::mode::read_write>(cgh);
+     typename bitonic_sort_base<T, 4>::local_buffer_accessor_t l(
+         range<1>(bitonic_sort_base<T, 4>::vec_elems_per_thread_::value *
+                  localWorkSize),
+         cgh);
+
+     cgh.parallel_for<kernel_bitonic_sort_init<T, Alloc>>(
+         ndrange, bitonic_sort_init<T, 4>(g, l));
+   })
+      .wait();
+
+  q.wait_and_throw();
+  return;
```

Comment:
why is this return here? there is code afterwards

Reply:
debugging artifact.